File: sdk/python/backstage_idp_sdk/client/backstage_client.py
"""Main Backstage client for Python SDK."""

from typing import Any, Dict, List, Optional, Union, Callable, AsyncGenerator
import asyncio
from datetime import datetime

from ..types import (
    ClientConfig,
    AuthConfig,
    RequestOptions,
    QueryParams,
    HealthCheck,
    Plugin,
    PluginInstallRequest,
    PluginInstallResponse,
    Workflow,
    WorkflowRequest,
    WorkflowExecution,
    WorkflowExecutionRequest,
    Tenant,
    TenantRequest,
    TenantAnalytics,
    Notification,
    NotificationRequest,
    Metrics,
    SearchResponse,
    PaginatedResponse
)
from ..auth.auth_manager import AuthManager, TokenInfo
from .http_client import HttpClient
from ..graphql.graphql_client import GraphQLClient
from ..websocket.websocket_client import WebSocketClient
from ..exceptions import BackstageSDKError, ConfigurationError


class SystemAPI:
    """System health and status operations."""
    
    def __init__(self, http_client: HttpClient):
        self._http = http_client

    async def get_health(self, verbose: bool = False, options: Optional[RequestOptions] = None) -> HealthCheck:
        """Get system health status."""
        params = {"verbose": verbose} if verbose else {}
        response = await self._http.get("/health", params=params, options=options)
        return HealthCheck.model_validate(response)

    async def get_metrics(
        self, 
        timerange: str = "1h", 
        format: str = "json",
        options: Optional[RequestOptions] = None
    ) -> Metrics:
        """Get system metrics."""
        params = {"timerange": timerange, "format": format}
        response = await self._http.get("/metrics", params=params, options=options)
        return Metrics.model_validate(response)


class TenantsAPI:
    """Tenant management operations."""
    
    def __init__(self, http_client: HttpClient):
        self._http = http_client

    async def get_current(self, options: Optional[RequestOptions] = None) -> Dict[str, Any]:
        """Get current tenant information."""
        params = {"action": "current"}
        return await self._http.get("/tenants", params=params, options=options)

    async def list(self, options: Optional[RequestOptions] = None) -> Dict[str, List[Tenant]]:
        """List user's tenants."""
        params = {"action": "list"}
        response = await self._http.get("/tenants", params=params, options=options)
        return {
            "tenants": [Tenant.model_validate(t) for t in response["tenants"]]
        }

    async def get_analytics(self, days: int = 30, options: Optional[RequestOptions] = None) -> Dict[str, TenantAnalytics]:
        """Get tenant analytics."""
        params = {"action": "analytics", "days": days}
        response = await self._http.get("/tenants", params=params, options=options)
        return {
            "analytics": TenantAnalytics.model_validate(response["analytics"])
        }

    async def create(self, request: TenantRequest, options: Optional[RequestOptions] = None) -> Tenant:
        """Create new tenant."""
        response = await self._http.post("/tenants", json=request.model_dump(), options=options)
        return Tenant.model_validate(response)

    async def update(self, request: TenantRequest, options: Optional[RequestOptions] = None) -> Tenant:
        """Update tenant."""
        response = await self._http.post("/tenants", json=request.model_dump(), options=options)
        return Tenant.model_validate(response)


class PluginsAPI:
    """Plugin management operations."""
    
    def __init__(self, http_client: HttpClient):
        self._http = http_client

    async def list(
        self, 
        params: Optional[QueryParams] = None,
        options: Optional[RequestOptions] = None
    ) -> PaginatedResponse[Plugin]:
        """List available plugins."""
        query_params = params.model_dump(exclude_none=True) if params else {}
        response = await self._http.get("/plugins", params=query_params, options=options)
        
        return PaginatedResponse(
            items=[Plugin.model_validate(p) for p in response["plugins"]],
            total=response["total"],
            limit=response["limit"],
            offset=response["offset"],
            has_more=response["offset"] + len(response["plugins"]) < response["total"]
        )

    async def get(self, plugin_id: str, options: Optional[RequestOptions] = None) -> Plugin:
        """Get plugin details."""
        response = await self._http.get(f"/plugins/{plugin_id}", options=options)
        return Plugin.model_validate(response)

    async def install(
        self, 
        request: PluginInstallRequest,
        options: Optional[RequestOptions] = None
    ) -> PluginInstallResponse:
        """Install plugin."""
        response = await self._http.post("/plugins", json=request.model_dump(), options=options)
        return PluginInstallResponse.model_validate(response)

    async def uninstall(self, plugin_id: str, options: Optional[RequestOptions] = None) -> None:
        """Uninstall plugin."""
        await self._http.delete(f"/plugins/{plugin_id}", options=options)

    async def search(
        self, 
        query: str, 
        category: Optional[str] = None,
        options: Optional[RequestOptions] = None
    ) -> PaginatedResponse[Plugin]:
        """Search plugins."""
        params = QueryParams(search=query, category=category)
        return await self.list(params, options)


class WorkflowsAPI:
    """Workflow management operations."""
    
    def __init__(self, http_client: HttpClient):
        self._http = http_client

    async def list(
        self, 
        params: Optional[QueryParams] = None,
        options: Optional[RequestOptions] = None
    ) -> PaginatedResponse[Workflow]:
        """List workflows."""
        query_params = params.model_dump(exclude_none=True) if params else {}
        response = await self._http.get("/workflows", params=query_params, options=options)
        
        return PaginatedResponse(
            items=[Workflow.model_validate(w) for w in response["workflows"]],
            total=response["total"],
            limit=params.limit if params else 50,
            offset=params.offset if params else 0,
            has_more=(params.offset if params else 0) + len(response["workflows"]) < response["total"]
        )

    async def get(self, workflow_id: str, options: Optional[RequestOptions] = None) -> Workflow:
        """Get workflow details."""
        response = await self._http.get(f"/workflows/{workflow_id}", options=options)
        return Workflow.model_validate(response)

    async def create(self, request: WorkflowRequest, options: Optional[RequestOptions] = None) -> Workflow:
        """Create workflow."""
        response = await self._http.post("/workflows", json=request.model_dump(), options=options)
        return Workflow.model_validate(response)

    async def update(
        self, 
        workflow_id: str, 
        request: WorkflowRequest,
        options: Optional[RequestOptions] = None
    ) -> Workflow:
        """Update workflow."""
        response = await self._http.put(f"/workflows/{workflow_id}", json=request.model_dump(), options=options)
        return Workflow.model_validate(response)

    async def execute(
        self,
        workflow_id: str,
        request: Optional[WorkflowExecutionRequest] = None,
        options: Optional[RequestOptions] = None
    ) -> WorkflowExecution:
        """Execute workflow."""
        payload = request.model_dump() if request else {}
        response = await self._http.post(f"/workflows/{workflow_id}/execute", json=payload, options=options)
        return WorkflowExecution.model_validate(response)

    async def delete(self, workflow_id: str, options: Optional[RequestOptions] = None) -> None:
        """Delete workflow."""
        await self._http.delete(f"/workflows/{workflow_id}", options=options)


class NotificationsAPI:
    """Notification management operations."""
    
    def __init__(self, http_client: HttpClient):
        self._http = http_client

    async def list(
        self,
        unread_only: bool = False,
        limit: int = 20,
        options: Optional[RequestOptions] = None
    ) -> Dict[str, Any]:
        """List notifications."""
        params = {"unread_only": unread_only, "limit": limit}
        response = await self._http.get("/notifications", params=params, options=options)
        
        return {
            "notifications": [Notification.model_validate(n) for n in response["notifications"]],
            "total": response["total"],
            "unreadCount": response.get("unreadCount", 0)
        }

    async def create(
        self, 
        request: NotificationRequest,
        options: Optional[RequestOptions] = None
    ) -> Notification:
        """Create notification."""
        response = await self._http.post("/notifications", json=request.model_dump(), options=options)
        return Notification.model_validate(response)

    async def mark_as_read(self, notification_id: str, options: Optional[RequestOptions] = None) -> None:
        """Mark notification as read."""
        await self._http.patch(f"/notifications/{notification_id}", json={"read": True}, options=options)


class EventsAPI:
    """Event subscription operations."""
    
    def __init__(self, websocket_client: Optional[WebSocketClient]):
        self._ws = websocket_client

    async def on_plugin_event(self, handler: Callable[[Dict[str, Any]], None]) -> Optional[str]:
        """Subscribe to plugin events."""
        if not self._ws:
            return None
        return await self._ws.subscribe(
            ["plugin.installed", "plugin.uninstalled", "plugin.updated"],
            handler
        )

    async def on_workflow_event(self, handler: Callable[[Dict[str, Any]], None]) -> Optional[str]:
        """Subscribe to workflow events."""
        if not self._ws:
            return None
        return await self._ws.subscribe(
            ["workflow.started", "workflow.completed", "workflow.failed"],
            handler
        )

    async def on_system_event(self, handler: Callable[[Dict[str, Any]], None]) -> Optional[str]:
        """Subscribe to system events."""
        if not self._ws:
            return None
        return await self._ws.subscribe(["system.health_change"], handler)

    async def unsubscribe(self, subscription_id: str) -> None:
        """Unsubscribe from events."""
        if self._ws:
            await self._ws.unsubscribe(subscription_id)


class AuthAPI:
    """Authentication operations."""
    
    def __init__(self, auth_manager: Optional[AuthManager]):
        self._auth = auth_manager

    def set_api_key(self, api_key: str) -> None:
        """Set API key for authentication."""
        if self._auth:
            token_info = TokenInfo(access_token=api_key, token_type="api_key")
            self._auth.set_tokens(token_info)

    def set_bearer_token(self, token: str, refresh_token: Optional[str] = None) -> None:
        """Set bearer token for authentication.""" 
        if self._auth:
            token_info = TokenInfo(
                access_token=token,
                refresh_token=refresh_token,
                token_type="bearer",
                expires_at=self._auth._extract_expiry_from_jwt(token)
            )
            self._auth.set_tokens(token_info)

    def clear_auth(self) -> None:
        """Clear authentication tokens."""
        if self._auth:
            self._auth.clear_tokens()

    def is_authenticated(self) -> bool:
        """Check if authenticated."""
        return self._auth.is_authenticated() if self._auth else False

    def get_access_token(self) -> Optional[str]:
        """Get current access token."""
        return self._auth.get_access_token() if self._auth else None


class BackstageClient:
    """Main Backstage client."""
    
    def __init__(self, config: ClientConfig):
        """Initialize Backstage client."""
        self.config = config
        
        # Initialize auth manager
        auth_config = None
        if config.api_key or config.bearer_token:
            auth_config = AuthConfig(
                base_url=config.base_url,
                api_key=config.api_key,
                bearer_token=config.bearer_token,
                auto_refresh=True
            )
            self._auth_manager = AuthManager(auth_config)
        else:
            self._auth_manager = None
        
        # Initialize HTTP client
        self._http_client = HttpClient(config, self._auth_manager)
        
        # Initialize GraphQL client
        self._graphql_client = GraphQLClient(config, self._auth_manager)
        
        # Initialize WebSocket client
        self._websocket_client = WebSocketClient(config, self._auth_manager)
        
        # Initialize API clients
        self.system = SystemAPI(self._http_client)
        self.tenants = TenantsAPI(self._http_client)
        self.plugins = PluginsAPI(self._http_client)
        self.workflows = WorkflowsAPI(self._http_client)
        self.notifications = NotificationsAPI(self._http_client)
        self.events = EventsAPI(self._websocket_client)
        self.auth = AuthAPI(self._auth_manager)

    async def connect(self) -> None:
        """Connect WebSocket client."""
        if self._websocket_client:
            await self._websocket_client.connect()

    async def disconnect(self) -> None:
        """Disconnect WebSocket client."""
        if self._websocket_client:
            await self._websocket_client.disconnect()

    async def search(
        self,
        query: str,
        type: Optional[str] = None,
        limit: int = 20,
        options: Optional[RequestOptions] = None
    ) -> SearchResponse:
        """Global search."""
        params = {"q": query, "limit": limit}
        if type:
            params["type"] = type
            
        response = await self._http_client.get("/search", params=params, options=options)
        return SearchResponse.model_validate(response)

    async def batch_execute(
        self,
        operations: List[Dict[str, Any]],
        options: Optional[RequestOptions] = None
    ) -> List[Any]:
        """Execute multiple operations in batch."""
        payload = {"operations": operations}
        return await self._http_client.post("/batch", json=payload, options=options)

    def get_connection_status(self) -> Dict[str, Any]:
        """Get connection status."""
        return {
            "http": True,  # HTTP client doesn't maintain persistent connection
            "websocket": self._websocket_client.get_status() if self._websocket_client else None,
            "auth": self.auth.is_authenticated()
        }

    def get_statistics(self) -> Dict[str, Any]:
        """Get client statistics."""
        return {
            "subscriptions": self._websocket_client.get_subscriptions_count() if self._websocket_client else 0,
            "circuit_breaker": self._http_client.get_circuit_breaker_status()
        }

    @property
    def graphql(self) -> GraphQLClient:
        """Access GraphQL client."""
        return self._graphql_client

    @property
    def websocket(self) -> Optional[WebSocketClient]:
        """Access WebSocket client."""
        return self._websocket_client

    async def close(self) -> None:
        """Close all connections and cleanup resources."""
        await asyncio.gather(
            self._http_client.close(),
            self._graphql_client.close(),
            self._websocket_client.close() if self._websocket_client else asyncio.sleep(0),
            self._auth_manager.close() if self._auth_manager else asyncio.sleep(0),
            return_exceptions=True
        )

    async def __aenter__(self):
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.close()

File: sdk/typescript/src/client/backstage-client.ts
import { EventEmitter } from 'eventemitter3';

import { AuthManager, AuthConfig } from '../auth/auth-manager';
import { HttpClient, HttpClientConfig } from './http-client';
import { BackstageGraphQLClient, GraphQLClientConfig } from '../graphql/graphql-client';
import { BackstageWebSocketClient, WebSocketClientConfig } from '../websocket/websocket-client';
import { 
  ClientConfig,
  HealthCheck,
  Plugin,
  PluginInstallRequest,
  PluginInstallResponse,
  Workflow,
  WorkflowRequest,
  WorkflowExecution,
  WorkflowExecutionRequest,
  Tenant,
  TenantRequest,
  TenantAnalytics,
  Notification,
  NotificationRequest,
  Metrics,
  SearchResponse,
  RequestOptions,
  PaginatedResponse,
  QueryParams
} from '../types';

export class BackstageClient extends EventEmitter {
  private authManager: AuthManager;
  private httpClient: HttpClient;
  private graphqlClient: BackstageGraphQLClient;
  private wsClient: BackstageWebSocketClient;

  constructor(config: ClientConfig) {
    super();

    // Initialize auth manager
    const authConfig: AuthConfig = {
      baseURL: config.baseURL,
      apiKey: config.apiKey,
      bearerToken: config.bearerToken,
      autoRefresh: true,
    };
    this.authManager = new AuthManager(authConfig);

    // Initialize HTTP client
    const httpConfig: HttpClientConfig = {
      baseURL: config.baseURL,
      timeout: config.timeout,
      retries: config.retries,
      retryDelay: config.retryDelay,
      circuitBreakerOptions: config.circuitBreakerOptions,
      rateLimit: config.rateLimit,
    };
    this.httpClient = new HttpClient(httpConfig, this.authManager);

    // Initialize GraphQL client
    const graphqlConfig: GraphQLClientConfig = {
      endpoint: `${config.baseURL}/graphql`,
      wsEndpoint: `${config.baseURL.replace('http', 'ws')}/graphql`,
      timeout: config.timeout,
    };
    this.graphqlClient = new BackstageGraphQLClient(graphqlConfig, this.authManager);

    // Initialize WebSocket client
    const wsConfig: WebSocketClientConfig = {
      url: `${config.baseURL.replace('http', 'ws')}/ws`,
      reconnect: true,
      reconnectInterval: 5000,
      maxReconnectAttempts: 10,
    };
    this.wsClient = new BackstageWebSocketClient(wsConfig, this.authManager);

    this.setupEventForwarding();
  }

  /**
   * Setup event forwarding from sub-clients
   */
  private setupEventForwarding(): void {
    // Forward auth events
    this.authManager.on('tokensUpdated', (tokens) => this.emit('tokensUpdated', tokens));
    this.authManager.on('authError', (error) => this.emit('authError', error));

    // Forward HTTP client events
    this.httpClient.on('requestStart', (data) => this.emit('requestStart', data));
    this.httpClient.on('requestSuccess', (data) => this.emit('requestSuccess', data));
    this.httpClient.on('requestError', (data) => this.emit('requestError', data));

    // Forward WebSocket events
    this.wsClient.on('connected', () => this.emit('connected'));
    this.wsClient.on('disconnected', (data) => this.emit('disconnected', data));
    this.wsClient.on('event', (event) => this.emit('event', event));
  }

  /**
   * Connect WebSocket client
   */
  public async connect(): Promise<void> {
    return this.wsClient.connect();
  }

  /**
   * Disconnect WebSocket client
   */
  public disconnect(): void {
    this.wsClient.disconnect();
  }

  /**
   * Authentication methods
   */
  public auth = {
    /**
     * Set API key for authentication
     */
    setApiKey: (apiKey: string): void => {
      this.authManager.setTokens({
        accessToken: apiKey,
        tokenType: 'api-key',
      });
    },

    /**
     * Set bearer token for authentication
     */
    setBearerToken: (token: string, refreshToken?: string): void => {
      this.authManager.setTokens({
        accessToken: token,
        refreshToken,
        tokenType: 'bearer',
      });
    },

    /**
     * Clear authentication tokens
     */
    clearAuth: (): void => {
      this.authManager.clearTokens();
    },

    /**
     * Check if authenticated
     */
    isAuthenticated: (): boolean => {
      return this.authManager.isAuthenticated();
    },

    /**
     * Get current access token
     */
    getAccessToken: (): string | null => {
      return this.authManager.getAccessToken();
    },
  };

  /**
   * System health and status
   */
  public system = {
    /**
     * Get system health status
     */
    getHealth: async (verbose = false, options?: RequestOptions): Promise<HealthCheck> => {
      return this.httpClient.get<HealthCheck>('/health', { verbose }, options);
    },

    /**
     * Get system metrics
     */
    getMetrics: async (
      timerange: '1h' | '24h' | '7d' | '30d' = '1h',
      format: 'json' | 'prometheus' = 'json',
      options?: RequestOptions
    ): Promise<Metrics> => {
      return this.httpClient.get<Metrics>('/metrics', { timerange, format }, options);
    },
  };

  /**
   * Tenant management
   */
  public tenants = {
    /**
     * Get current tenant information
     */
    getCurrent: async (options?: RequestOptions): Promise<{ tenant: Tenant; userRole?: string; isOwner?: boolean }> => {
      return this.httpClient.get('/tenants', { action: 'current' }, options);
    },

    /**
     * List user's tenants
     */
    list: async (options?: RequestOptions): Promise<{ tenants: Tenant[] }> => {
      return this.httpClient.get('/tenants', { action: 'list' }, options);
    },

    /**
     * Get tenant analytics
     */
    getAnalytics: async (days = 30, options?: RequestOptions): Promise<{ analytics: TenantAnalytics }> => {
      return this.httpClient.get('/tenants', { action: 'analytics', days }, options);
    },

    /**
     * Create new tenant
     */
    create: async (request: TenantRequest, options?: RequestOptions): Promise<Tenant> => {
      return this.httpClient.post('/tenants', request, options);
    },

    /**
     * Update tenant
     */
    update: async (request: TenantRequest, options?: RequestOptions): Promise<Tenant> => {
      return this.httpClient.post('/tenants', request, options);
    },
  };

  /**
   * Plugin management
   */
  public plugins = {
    /**
     * List available plugins
     */
    list: async (params: QueryParams = {}, options?: RequestOptions): Promise<PaginatedResponse<Plugin>> => {
      const response = await this.httpClient.get<{
        plugins: Plugin[];
        total: number;
        limit: number;
        offset: number;
      }>('/plugins', params, options);
      
      return {
        items: response.plugins,
        total: response.total,
        limit: response.limit,
        offset: response.offset,
        hasMore: response.offset + response.plugins.length < response.total,
      };
    },

    /**
     * Get plugin details
     */
    get: async (pluginId: string, options?: RequestOptions): Promise<Plugin> => {
      return this.httpClient.get<Plugin>(`/plugins/${pluginId}`, undefined, options);
    },

    /**
     * Install plugin
     */
    install: async (request: PluginInstallRequest, options?: RequestOptions): Promise<PluginInstallResponse> => {
      return this.httpClient.post<PluginInstallResponse>('/plugins', request, options);
    },

    /**
     * Uninstall plugin
     */
    uninstall: async (pluginId: string, options?: RequestOptions): Promise<void> => {
      return this.httpClient.delete(`/plugins/${pluginId}`, options);
    },

    /**
     * Search plugins
     */
    search: async (query: string, category?: string, options?: RequestOptions): Promise<PaginatedResponse<Plugin>> => {
      return this.plugins.list({ search: query, category }, options);
    },
  };

  /**
   * Workflow management
   */
  public workflows = {
    /**
     * List workflows
     */
    list: async (params: QueryParams = {}, options?: RequestOptions): Promise<PaginatedResponse<Workflow>> => {
      const response = await this.httpClient.get<{
        workflows: Workflow[];
        total: number;
      }>('/workflows', params, options);
      
      return {
        items: response.workflows,
        total: response.total,
        limit: params.limit || 50,
        offset: params.offset || 0,
        hasMore: (params.offset || 0) + response.workflows.length < response.total,
      };
    },

    /**
     * Get workflow details
     */
    get: async (workflowId: string, options?: RequestOptions): Promise<Workflow> => {
      return this.httpClient.get<Workflow>(`/workflows/${workflowId}`, undefined, options);
    },

    /**
     * Create workflow
     */
    create: async (request: WorkflowRequest, options?: RequestOptions): Promise<Workflow> => {
      return this.httpClient.post<Workflow>('/workflows', request, options);
    },

    /**
     * Update workflow
     */
    update: async (workflowId: string, request: WorkflowRequest, options?: RequestOptions): Promise<Workflow> => {
      return this.httpClient.put<Workflow>(`/workflows/${workflowId}`, request, options);
    },

    /**
     * Execute workflow
     */
    execute: async (
      workflowId: string,
      request: WorkflowExecutionRequest = {},
      options?: RequestOptions
    ): Promise<WorkflowExecution> => {
      return this.httpClient.post<WorkflowExecution>(`/workflows/${workflowId}/execute`, request, options);
    },

    /**
     * Delete workflow
     */
    delete: async (workflowId: string, options?: RequestOptions): Promise<void> => {
      return this.httpClient.delete(`/workflows/${workflowId}`, options);
    },
  };

  /**
   * Notification management
   */
  public notifications = {
    /**
     * List notifications
     */
    list: async (unreadOnly = false, limit = 20, options?: RequestOptions): Promise<{
      notifications: Notification[];
      total: number;
      unreadCount: number;
    }> => {
      return this.httpClient.get('/notifications', { unread_only: unreadOnly, limit }, options);
    },

    /**
     * Create notification
     */
    create: async (request: NotificationRequest, options?: RequestOptions): Promise<Notification> => {
      return this.httpClient.post<Notification>('/notifications', request, options);
    },

    /**
     * Mark notification as read
     */
    markAsRead: async (notificationId: string, options?: RequestOptions): Promise<void> => {
      return this.httpClient.patch(`/notifications/${notificationId}`, { read: true }, options);
    },
  };

  /**
   * Search functionality
   */
  public search = async (
    query: string,
    type?: 'plugin' | 'workflow' | 'service' | 'documentation',
    limit = 20,
    options?: RequestOptions
  ): Promise<SearchResponse> => {
    return this.httpClient.get<SearchResponse>('/search', { q: query, type, limit }, options);
  };

  /**
   * GraphQL client access
   */
  public graphql = this.graphqlClient;

  /**
   * WebSocket client access
   */
  public websocket = this.wsClient;

  /**
   * Event subscription methods
   */
  public events = {
    /**
     * Subscribe to plugin events
     */
    onPluginEvent: (handler: (event: any) => void) => {
      return this.wsClient.subscribe(['plugin.installed', 'plugin.uninstalled', 'plugin.updated'], handler);
    },

    /**
     * Subscribe to workflow events
     */
    onWorkflowEvent: (handler: (event: any) => void) => {
      return this.wsClient.subscribe(['workflow.started', 'workflow.completed', 'workflow.failed'], handler);
    },

    /**
     * Subscribe to system events
     */
    onSystemEvent: (handler: (event: any) => void) => {
      return this.wsClient.subscribe(['system.health_change'], handler);
    },

    /**
     * Subscribe to all events
     */
    onAnyEvent: (handler: (event: any) => void) => {
      this.wsClient.on('event', handler);
      return () => this.wsClient.off('event', handler);
    },
  };

  /**
   * Batch operations
   */
  public batch = {
    /**
     * Execute multiple operations in a single request
     */
    execute: async (operations: Array<{
      method: 'GET' | 'POST' | 'PUT' | 'DELETE';
      path: string;
      data?: any;
    }>, options?: RequestOptions) => {
      return this.httpClient.post('/batch', { operations }, options);
    },
  };

  /**
   * Utility methods
   */
  public utils = {
    /**
     * Get connection status
     */
    getConnectionStatus: () => ({
      http: true, // HTTP client doesn't maintain persistent connection
      websocket: this.wsClient.getStatus(),
      auth: this.authManager.isAuthenticated(),
    }),

    /**
     * Get client statistics
     */
    getStatistics: () => ({
      subscriptions: this.wsClient.getSubscriptionsCount(),
      circuitBreaker: this.httpClient.getCircuitBreakerStatus(),
    }),
  };

  /**
   * Dispose of the client and clean up resources
   */
  public dispose(): void {
    this.authManager.dispose();
    this.httpClient.dispose();
    this.graphqlClient.dispose();
    this.wsClient.dispose();
    this.removeAllListeners();
  }
}